### PR TITLE
Support computing the digest when caller can only specify --tarball.

### PR DESCRIPTION
**Pull request by faximan**
This was supported before, but was removed as part of https://github.com/bazelbuild/rules_docker/pull/1123, a performance optimization/refactoring change. I am proposing to add back the support without changing the code path if the other inputs are available, so current users are not affected.

### CODE REVIEW

**Review by smukherj1 — APPROVED** [review]

**Comment by faximan** [comment]
Seems like one of the automatic build failed

```
(13:45:48) ERROR: /var/lib/buildkite-agent/.cache/bazel/_bazel_buildkite-agent/ec321eb2cc2d0f8f91b676b6d4c66c29/external/go_image_base/image/BUILD:4:17: SHA256 external/go_image_base/image/000.tar.gz.nogz.sha256 failed: (Exit 34): java.io.IOException: io.grpc.StatusRuntimeException: UNAVAILABLE: HTTP/2 error code: NO_ERROR
--
  | Received Goaway
<snip>
(15:39:41) ERROR: The Build Event Protocol local file upload failed: findMissingBlobs(3) for bes-upload: UNAVAILABLE: HTTP/2 error code: NO_ERROR
--
  | Received Goaway
  | load_shed
```

No idea what this means, but looks unrelated?
https://buildkite.com/bazel/rules-docker-docker/builds/4914#420b3b44-de3b-434f-9de4-946cd2b5479a

**Comment by smukherj1** [comment]
CI is broken because of https://github.com/bazelbuild/bazel/issues/12920 :(

This is blocked until that is resolved

**Comment by faximan** [comment]
Breakage seems resolved. How do we move forward? (if I can trigger the merge I didn't figure out how)

**Comment by smukherj1** [comment]
There's still the following failure:

```
Extracting Bazel installation...
Starting local Bazel server and connecting to it...
Loading: 
Loading: 0 packages loaded
Loading: 0 packages loaded
Analyzing: 9 targets (1 packages loaded, 0 targets configured)
Analyzing: 9 targets (9 packages loaded, 6 targets configured)
Analyzing: 9 targets (9 packages loaded, 6 targets configured)
Analyzing: 9 targets (9 packages loaded, 6 targets configured)
Analyzing: 9 targets (44 packages loaded, 6864 targets configured)
Analyzing: 9 targets (54 packages loaded, 6899 targets configured)
Analyzing: 9 targets (60 packages loaded, 6978 targets configured)
Analyzing: 9 targets (61 packages loaded, 7007 targets configured)
DEBUG: Rule 'bazel_image' indicated that a canonical reproducible form can be obtained by modifying arguments digest = "sha256:ace9881e6e9c5d48b5fd637321361aeffe54000265894a65f7d818dc1065bd80"
DEBUG: Repository bazel_image instantiated at:
  no stack (--record_rule_instantiation_callstack not enabled)
Repository rule container_pull defined at:
  /builder/home/.cache/bazel/_bazel_root/49800fdee5413a0f3c3f662f375d96d7/external/io_bazel_rules_docker/container/pull.bzl:243:33: in <toplevel>
Analyzing: 9 targets (62 packages loaded, 7014 targets configured)
Analyzing: 9 targets (63 packages loaded, 7135 targets configured)
Analyzing: 9 targets (82 packages loaded, 7258 targets configured)
Analyzing: 9 targets (82 packages loaded, 7258 targets configured)
Analyzing: 9 targets (82 packages loaded, 7258 targets configured)
Analyzing: 9 targets (82 packages loaded, 7258 targets configured)
INFO: Analyzed 9 targets (83 packages loaded, 7262 targets configured).
INFO: Found 7 targets and 2 test targets...
[0 / 100] [Prepa] Expanding template external/io_bazel_rules_docker/container/build_tar [for host] ... (6 actions, 0 running)
[105 / 115] ExtractConfig run_instruction_arbitrary/bazel_gcloud_dockerfile_wrapper.dockerfile_image.tar.config; 7s linux-sandbox ... (2 actions, 1 running)
[106 / 115] ExtractConfig run_instruction_arbitrary/bazel_gcloud_dockerfile_wrapper.dockerfile_image.tar.config; 26s linux-sandbox ... (2 actions running)
[107 / 115] Action run_instruction_arbitrary/gcloud_install_commit.tar; 32s linux-sandbox ... (2 actions running)
FAIL: //run_instruction_arbitrary:bazel_gcloud_dockerfile_test (see /builder/home/.cache/bazel/_bazel_root/49800fdee5413a0f3c3f662f375d96d7/execroot/dockerfile_examples/bazel-out/k8-fastbuild/testlogs/run_instruction_arbitrary/bazel_gcloud_dockerfile_test/test.log)
[111 / 116] 1 / 2 tests, 1 failed; Action run_instruction_arbitrary/gcloud_install_commit.tar; 55s linux-sandbox
[111 / 116] 1 / 2 tests, 1 failed; Action run_instruction_arbitrary/gcloud_install_commit.tar; 81s linux-sandbox
INFO: From Action run_instruction_arbitrary/gcloud_install_commit.tar:
Loaded image: bazel/run_instruction_arbitrary:gcloud_installer
sha256:f99309a5f3cfbf0d241b747e9a051df421acf1000cd05a560d526403a7f26afe
3267410fdfd1629fb8ea4fedfdbbacb1ca39fc6f70d0802e8f16e9faca2339b5
[112 / 116] 1 / 2 tests, 1 failed; [Prepa] ExtractConfig run_instruction_arbitrary/bazel_gcloud_bazel_wrapper.gcloud_install_commit.tar.config
[116 / 117] 1 / 2 tests, 1 failed; Testing //run_instruction_arbitrary:bazel_gcloud_bazel_test; 11s linux-sandbox
FAIL: //run_instruction_arbitrary:bazel_gcloud_bazel_test (see /builder/home/.cache/bazel/_bazel_root/49800fdee5413a0f3c3f662f375d96d7/execroot/dockerfile_examples/bazel-out/k8-fastbuild/testlogs/run_instruction_arbitrary/bazel_gcloud_bazel_test/test.log)
INFO: Elapsed time: 302.623s, Critical Path: 199.65s
INFO: 69 processes: 69 linux-sandbox.
INFO: Build completed, 2 tests FAILED, 117 total actions
//run_instruction_arbitrary:bazel_gcloud_bazel_test                      FAILED in 11.4s
  /builder/home/.cache/bazel/_bazel_root/49800fdee5413a0f3c3f662f375d96d7/execroot/dockerfile_examples/bazel-out/k8-fastbuild/testlogs/run_instruction_arbitrary/bazel_gcloud_bazel_test/test.log
//run_instruction_arbitrary:bazel_gcloud_dockerfile_test                 FAILED in 11.9s
  /builder/home/.cache/bazel/_bazel_root/49800fdee5413a0f3c3f662f375d96d7/execroot/dockerfile_examples/bazel-out/k8-fastbuild/testlogs/run_instruction_arbitrary/bazel_gcloud_dockerfile_test/test.log

Executed 2 out of 2 tests: 2 fail locally.
INFO: Build completed, 2 tests FAILED, 117 total actions
```

You might be able to trigger it by running `bazel test //run_instruction_arbitrary:bazel_gcloud_bazel_test`

**Comment by faximan** [comment]
That test also fails without my change with

```
=== RUN: Command Test: pip_test
--- FAIL
duration: 753.336685ms
stderr: Traceback (most recent call last):
  File "/usr/local/bin/pip", line 7, in <module>
    from pip._internal.cli.main import main
  File "/usr/local/lib/python3.5/dist-packages/pip/_internal/cli/main.py", line 60
    sys.stderr.write(f"ERROR: {exc}")
                                   ^
SyntaxError: invalid syntax

Error: Expected string 'pip \d+.*' not found in output ''
Error: Test 'pip_test' exited with incorrect error code. Expected: 0, Actual: 1
```

Note: I don't have permissions to check the cloud build link from the report here (https://console.cloud.google.com/cloud-build/builds/e5d40bc1-2a75-409c-aab3-ef6a1978c2a6?project=1026272623418) to verify that the merge tests failed with the same error.

If it weren't for the "python3.5" path in the test log, I'd have thought this is because pip dropped support for python2 recently (https://stackoverflow.com/questions/65896334/python-pip-broken-wiith-sys-stderr-writeferror-exc/65928602), but now I am not sure.

In any case, this seems completely unrelated to my change, so I'd be grateful if one of the maintainers can take a look. Thanks!

**Comment by faximan** [comment]
Ok, attempted a fix to the broken tests in https://github.com/bazelbuild/rules_docker/pull/1723.
You probably have a better idea how to do it properly but that should give you some clues what the problem is.

**Comment by smukherj1** [comment]
@faximan could you rebase on master?

**Pull request merged**
